ls1intum/Artemis
# [Exam] Test Run follow-up

# Checklist

### Motivation and Context
As a follow-up to #2096 this PR addresses:
- REST API: Change names of rest call for delete & create test run 
- Currently, the test run exercises are ordered by the `id` of the exercise groups. They should follow the order of the exercises groups, as configured for the exam.
- Empty exercise groups do not allow the creation of test runs. They should be ignored when creating a test run. 
- Reduce Code duplication
- Improve the performance, complexity and readability of `studentExamService.deleteTestRun()`

**Model Changes:**
- Add `started_date` to `student_exam`. This should be set only when the student first starts the exam conduction.
- Change `last_modified_date`, `created_date` and `submission_date` datatype of `student_exam` to `TIMESTAMP`
- Change `last_modified_date` `created_date` datatype of `exam_session` to `TIMESTAMP`


### Description

- Change Routes to:
> DELETE courses/{courseId}/exams/{examId}/test-run/{testRunId}
> POST courses/{courseId}/exams/{examId}/test-run/
- Lock the create test run button if no exercises are found
- Ignore exercise groups with no exercises in the test run modal
- Keep the order of the exercise groups for the test runs
- Refactor the deleteTestRun() to reduce complexity and improve performance
> Fetch only relevant exercise participations (no need to filter after)
- Improve the client test and server test coverage

### Steps for Testing
1. Log in to Artemis
2. Create an exam with 2 exercise groups and no exercises
3. Navigate to the test runs, you should see the locked button and the tooltip for `Create Test Run` 
4. Add an exercise to one of the exercise groups
5. Navigate to the test runs, create a test run with one exercise.
6. Start the test run and finish it
7. Add an exercise to the other exercise group
8. Change the order of the exercise groups
9. Create a new test run. 
10. Check the order of the exercises, it should follow the order set for the exercise groups

### Test Coverage
**Client:**

- create-test-run-modal.component.ts 64% -> 70.59%
- test-run-management.component.ts 77% -> 79.03%

**Server:**

- StudentExamService: 75% -> 82%


### Screenshots
Added Tooltips when disabling buttons: 
![image](https://user-images.githubusercontent.com/15110960/94366868-1b40c700-00db-11eb-9df2-466f8a264d14.png)
![image](https://user-images.githubusercontent.com/15110960/94366927-812d4e80-00db-11eb-954e-51094c0abf34.png)
